tensorflow/python/keras/engine/data_adapter.py

```diff
@@ -1412,7 +1412,30 @@ def _make_class_weight_map_fn(class_weight):
     A function that can be used with `tf.data.Dataset.map` to apply class
     weighting.
   """
-  class_ids = list(sorted(class_weight.keys()))
+  # Validate that all keys are integers
+  for key in class_weight.keys():
+    if not isinstance(key, (int, numpy_compat.integer_types)):
+      try:
+        # Try to convert to int if it's a string representation
+        if isinstance(key, str):
+          int(key)
+        else:
+          raise ValueError()
+      except (ValueError, TypeError):
+        raise ValueError(f"Invalid class_weight key: '{key}'. "
+                        f"Class weight keys must be integers representing "
+                        f"class indices,"
+                        f"got key of type {type(key).__name__}.")
```

Comment:
Why not just convert directly to the key used in the `converted_class_weight` below? Right now you are parsing the string twice.